File: src/ffi_helpers.js
const _sh_ptr_read_c_char = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_char(ptr: c_ptr, offset: c_int): c_char { throw 0; });
const _sh_ptr_write_c_char = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_char(ptr: c_ptr, offset: c_int, v: c_char): void {});
const _sh_ptr_read_c_schar = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_schar(ptr: c_ptr, offset: c_int): c_schar { throw 0; });
const _sh_ptr_write_c_schar = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_schar(ptr: c_ptr, offset: c_int, v: c_schar): void {});
const _sh_ptr_read_c_uchar = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_uchar(ptr: c_ptr, offset: c_int): c_uchar { throw 0; });
const _sh_ptr_write_c_uchar = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_uchar(ptr: c_ptr, offset: c_int, v: c_uchar): void {});
const _sh_ptr_read_c_short = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_short(ptr: c_ptr, offset: c_int): c_short { throw 0; });
const _sh_ptr_write_c_short = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_short(ptr: c_ptr, offset: c_int, v: c_short): void {});
const _sh_ptr_read_c_ushort = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_ushort(ptr: c_ptr, offset: c_int): c_ushort { throw 0; });
const _sh_ptr_write_c_ushort = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_ushort(ptr: c_ptr, offset: c_int, v: c_ushort): void {});
const _sh_ptr_read_c_int = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_int(ptr: c_ptr, offset: c_int): c_int { throw 0; });
const _sh_ptr_write_c_int = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_int(ptr: c_ptr, offset: c_int, v: c_int): void {});
const _sh_ptr_read_c_uint = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_uint(ptr: c_ptr, offset: c_int): c_uint { throw 0; });
const _sh_ptr_write_c_uint = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_uint(ptr: c_ptr, offset: c_int, v: c_uint): void {});
const _sh_ptr_read_c_long = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_long(ptr: c_ptr, offset: c_int): c_long { throw 0; });
const _sh_ptr_write_c_long = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_long(ptr: c_ptr, offset: c_int, v: c_long): void {});
const _sh_ptr_read_c_ulong = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_ulong(ptr: c_ptr, offset: c_int): c_ulong { throw 0; });
const _sh_ptr_write_c_ulong = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_ulong(ptr: c_ptr, offset: c_int, v: c_ulong): void {});
const _sh_ptr_read_c_longlong = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_longlong(ptr: c_ptr, offset: c_int): c_longlong { throw 0; });
const _sh_ptr_write_c_longlong = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_longlong(ptr: c_ptr, offset: c_int, v: c_longlong): void {});
const _sh_ptr_read_c_ulonglong = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_ulonglong(ptr: c_ptr, offset: c_int): c_ulonglong { throw 0; });
const _sh_ptr_write_c_ulonglong = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_ulonglong(ptr: c_ptr, offset: c_int, v: c_ulonglong): void {});
const _sh_ptr_read_c_bool = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_bool(ptr: c_ptr, offset: c_int): c_bool { throw 0; });
const _sh_ptr_write_c_bool = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_bool(ptr: c_ptr, offset: c_int, v: c_bool): void {});
const _sh_ptr_read_c_float = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_float(ptr: c_ptr, offset: c_int): c_float { throw 0; });
const _sh_ptr_write_c_float = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_float(ptr: c_ptr, offset: c_int, v: c_float): void {});
const _sh_ptr_read_c_double = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_double(ptr: c_ptr, offset: c_int): c_double { throw 0; });
const _sh_ptr_write_c_double = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_double(ptr: c_ptr, offset: c_int, v: c_double): void {});
const _sh_ptr_read_c_ptr = $SHBuiltin.extern_c({}, function _sh_ptr_read_c_ptr(ptr: c_ptr, offset: c_int): c_ptr { throw 0; });
const _sh_ptr_write_c_ptr = $SHBuiltin.extern_c({}, function _sh_ptr_write_c_ptr(ptr: c_ptr, offset: c_int, v: c_ptr): void {});
const _sh_ptr_add = $SHBuiltin.extern_c({}, function _sh_ptr_add(ptr: c_ptr, offset: c_int): c_ptr { throw 0; });

const _malloc = $SHBuiltin.extern_c({include: "stdlib.h"}, function malloc(size: c_size_t): c_ptr {
    throw 0;
});
const _free = $SHBuiltin.extern_c({include: "stdlib.h"}, function free(p: c_ptr): void {
});

File: src/ffi_helpers.c
#include <stdbool.h>

char _sh_ptr_read_c_char(char *ptr, int offset) {
  return *(char *)(ptr + offset);
}
void _sh_ptr_write_c_char(char *ptr, int offset, char v) {
  *(char *)(ptr + offset) = v;
}
signed char _sh_ptr_read_c_schar(char *ptr, int offset) {
  return *(signed char *)(ptr + offset);
}
void _sh_ptr_write_c_schar(char *ptr, int offset, signed char v) {
  *(signed char *)(ptr + offset) = v;
}
unsigned char _sh_ptr_read_c_uchar(char *ptr, int offset) {
  return *(unsigned char *)(ptr + offset);
}
void _sh_ptr_write_c_uchar(char *ptr, int offset, unsigned char v) {
  *(unsigned char *)(ptr + offset) = v;
}
short int _sh_ptr_read_c_short(char *ptr, int offset) {
  return *(short int *)(ptr + offset);
}
void _sh_ptr_write_c_short(char *ptr, int offset, short int v) {
  *(short int *)(ptr + offset) = v;
}
short unsigned int _sh_ptr_read_c_ushort(char *ptr, int offset) {
  return *(short unsigned int *)(ptr + offset);
}
void _sh_ptr_write_c_ushort(char *ptr, int offset, short unsigned int v) {
  *(short unsigned int *)(ptr + offset) = v;
}
int _sh_ptr_read_c_int(char *ptr, int offset) {
  return *(int *)(ptr + offset);
}
void _sh_ptr_write_c_int(char *ptr, int offset, int v) {
  *(int *)(ptr + offset) = v;
}
unsigned int _sh_ptr_read_c_uint(char *ptr, int offset) {
  return *(unsigned int *)(ptr + offset);
}
void _sh_ptr_write_c_uint(char *ptr, int offset, unsigned int v) {
  *(unsigned int *)(ptr + offset) = v;
}
long int _sh_ptr_read_c_long(char *ptr, int offset) {
  return *(long int *)(ptr + offset);
}
void _sh_ptr_write_c_long(char *ptr, int offset, long int v) {
  *(long int *)(ptr + offset) = v;
}
long unsigned int _sh_ptr_read_c_ulong(char *ptr, int offset) {
  return *(long unsigned int *)(ptr + offset);
}
void _sh_ptr_write_c_ulong(char *ptr, int offset, long unsigned int v) {
  *(long unsigned int *)(ptr + offset) = v;
}
long long int _sh_ptr_read_c_longlong(char *ptr, int offset) {
  return *(long long int *)(ptr + offset);
}
void _sh_ptr_write_c_longlong(char *ptr, int offset, long long int v) {
  *(long long int *)(ptr + offset) = v;
}
long long unsigned int _sh_ptr_read_c_ulonglong(char *ptr, int offset) {
  return *(long long unsigned int *)(ptr + offset);
}
void _sh_ptr_write_c_ulonglong(char *ptr, int offset, long long unsigned int v) {
  *(long long unsigned int *)(ptr + offset) = v;
}
bool _sh_ptr_read_c_bool(char *ptr, int offset) {
  return *(bool *)(ptr + offset);
}
void _sh_ptr_write_c_bool(char *ptr, int offset, bool v) {
  *(bool *)(ptr + offset) = v;
}
float _sh_ptr_read_c_float(char *ptr, int offset) {
  return *(float *)(ptr + offset);
}
void _sh_ptr_write_c_float(char *ptr, int offset, float v) {
  *(float *)(ptr + offset) = v;
}
double _sh_ptr_read_c_double(char *ptr, int offset) {
  return *(double *)(ptr + offset);
}
void _sh_ptr_write_c_double(char *ptr, int offset, double v) {
  *(double *)(ptr + offset) = v;
}
char *_sh_ptr_read_c_ptr(char *ptr, int offset) {
  return *(char **)(ptr + offset);
}
void _sh_ptr_write_c_ptr(char *ptr, int offset, char *v) {
  *(char **)(ptr + offset) = v;
}
char *_sh_ptr_add(char *ptr, int offset) {
  return ptr + offset;
}
